sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/aio/_file_system_client_async.py

```diff
@@ -663,6 +670,41 @@ async def delete_file(self, file,  # type: Union[FileProperties, str]
         await file_client.delete_file(**kwargs)
         return file_client
 
+    @distributed_trace_async
+    async def undelete_path(self, deleted_path_name, deleted_path_version, **kwargs):
+        # type: (str, str, **Any) -> Union[DataLakeDirectoryClient, DataLakeFileClient]
+        """Restores soft-deleted path.
+
+        Operation will only be successful if used within the specified number of days
+        set in the delete retention policy.
+
+        .. versionadded:: 12.3.0
+            This operation was introduced in API version '2020-06-12'.
+
+        :param str deleted_path_name:
+            Specifies the name of the deleted container to restore.
+        :param str deleted_path_version:
+            Specifies the version of the deleted container to restore.
+        :keyword int timeout:
+            The timeout parameter is expressed in seconds.
+        :rtype: ~azure.storage.blob.ContainerClient
+        """
+        _, url, undelete_source = self._undelete_path(deleted_path_name, deleted_path_version)
+
+        pipeline = AsyncPipeline(
+            transport=AsyncTransportWrapper(self._pipeline._transport), # pylint: disable = protected-access
+            policies=self._pipeline._impl_policies # pylint: disable = protected-access
+        )
+        path_client = DataLakeStorageClient(url, self.file_system_name, deleted_path_name, pipeline=pipeline)
+        try:
+            await path_client.path.undelete(undelete_source=undelete_source, **kwargs)
+        except StorageErrorException as error:
+            process_storage_error(error)
```

Comment:
later we need to get into this line and see if process_storage_error can parse the error. the error thrown by dfs endpoint is different blob endpoint, and this process_storage_error is for dfs endpoint